stratum/hal/lib/phal/tai/tai_optics_datasource.cc

```diff
@@ -0,0 +1,129 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "stratum/hal/lib/phal/tai/tai_optics_datasource.h"
+
+#include <cmath>
+
+#include "absl/memory/memory.h"
+#include "stratum/glue/integral_types.h"
+#include "stratum/glue/status/status.h"
+#include "stratum/glue/status/statusor.h"
+#include "stratum/hal/lib/common/common.pb.h"
+#include "stratum/hal/lib/phal/datasource.h"
+#include "stratum/hal/lib/phal/phal.pb.h"
+#include "stratum/lib/macros.h"
+
+namespace stratum {
+namespace hal {
+namespace phal {
+namespace tai {
+
+/*!
+ * \brief TaiOpticsDataSource::Make method \return constructed 
+ * TaiOpticsDataSource object
+ * TaiOpticsDataSource don't take ownership of \param tai_manager
+ *
+ * \note Ownership is transferred to the caller
+ */
+::util::StatusOr<std::shared_ptr<TaiOpticsDataSource>>
+TaiOpticsDataSource::Make(int id, tai::TAIManager* tai_manager,
+                          const PhalOpticalCardConfig& config) {
+  ASSIGN_OR_RETURN(auto cache, CachePolicyFactory::CreateInstance(
+                                   config.cache_policy().type(),
+                                   config.cache_policy().timed_value()));
+
+  auto optics_datasource = std::shared_ptr<TaiOpticsDataSource>(
+      new TaiOpticsDataSource(id, tai_manager, cache));
+
+  optics_datasource->UpdateValuesUnsafelyWithoutCacheOrLock();
+
+  return optics_datasource;
+}
+
+TaiOpticsDataSource::TaiOpticsDataSource(int id, tai::TAIManager* tai_manager,
+                                         CachePolicy* cache_policy)
+    : DataSource(cache_policy), tai_manager_(ABSL_DIE_IF_NULL(tai_manager)) {
+  // These values do not change during the lifetime of the data source.
+  module_slot_.AssignValue(id);
+
+  tx_laser_frequency_.AddSetter(
+      [this](uint64 requeste_tx_laser_frequency) -> ::util::Status {
+        return this->SetTxLaserFrequency(requeste_tx_laser_frequency);
+      });
+  modulation_format_.AddSetter(
+      [this](uint64 requeste_modulation_format) -> ::util::Status {
+        return this->SetModulationFormat(requeste_modulation_format);
+      });
+  output_power_.AddSetter(
+      [this](float requeste_output_power) -> ::util::Status {
+        return this->SetOutputPower(requeste_output_power);
+      });
+}
+
+/*!
+ * \brief TaiOpticsDataSource::UpdateValues method updates attributes with 
+ * fresh values from TAI.
+ * \return ::util::OkStatus if success
+ */
+::util::Status TaiOpticsDataSource::UpdateValues() {
+  ASSIGN_OR_RETURN(auto tx_laser_frequency,
+                   tai_manager_->GetValue<uint64>(
+                       TAI_NETWORK_INTERFACE_ATTR_TX_LASER_FREQ, {0, 0}));
```

Comment:
The module/interface pare are hardcoded. Should pass the module/interface-id when we create this data source. Same as below